Include/internal/pycore_runtime.h

```diff
@@ -84,6 +84,19 @@ typedef struct pyruntimestate {
        to access it, don't access it directly. */
     _Py_atomic_address _finalizing;
 
+    /* Tracks the finalize blocks.
+
+       Bit 0 is set to 1 by `Py_FinalizeEx` to indicate it is waiting to set `_finalizing`.
+
+       The remaining bits are a count of the number of finalize blocks that are
+       currently held.  Once bit 0 is set to 1, the number of finalize blocks is
+       not allowed to increase.
+
+       Protected by `ceval.gil.mutex`, and `ceval.gil.cond` must be broadcast
```

Comment:
We have a per-interpreter GIL now but - I think what needed to happen is to switch to using the main interpreters GIL as protection for this.  So that's what I've committed and pushed.

Alternatively this finalize_blocks could use a lock of its own?

Reply:
This PR seems like it should apply to each interpreter, rather than just to the whole global runtime (i.e. main interpreter).